Preserve h5py.Dataset filter settings on export

oruebel opened the pull request:
## Motivation

Fix #152 Preserve h5py.Dataset filter settings on export

## How to test the behavior?

- [x] Add unit tests 
- [X] Add functions to `ZarrDataIO` to wrap `h5py.Dataset` and infer filter settings
- [X] Update `ZarrIO` to wrap `h5py.Dataset` on write to preserve filter settings 

## Checklist

- [x] Did you update CHANGELOG.md with your changes?
- [X] Have you checked our [Contributing](https://github.com/hdmf-dev/hdmf-zarr/blob/dev/docs/CONTRIBUTING.rst) document?
- [X] Have you ensured the PR clearly describes the problem and the solution?
- [X] Is your contribution compliant with our coding style? This can be checked running `ruff` from the source directory.
- [X] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/hdmf-dev/hdmf-zarr/pulls) for the same change?
- [X] Have you included the relevant issue number using "Fix #XXX" notation where XXX is the issue number? By including "Fix #XXX" you allow GitHub to close issue #XXX when the PR is merged.

oruebel committed:
Fix #152 Preserve h5py.Dataset filter settings on export

oruebel committed:
Remove extra arg

oruebel committed:
Fix bad variable name

oruebel committed:
Add missing return statement in ZarrDataIO.from_h5py_dataset

oruebel committed:
Fix ruff

oruebel committed:
Add unit test for hdf5 to zarr export with filters

oruebel committed:
Update warning check in failing unit test

oruebel committed:
Fix bytes fill value

oruebel committed:
Fix sphinx docs warning

oruebel committed:
Fix docstring

oruebel committed:
Update Changelog

oruebel committed:
Fix ruff

oruebel commented:
@sinha-r @rly this PR is related to the cloud project to allow conversion from HDF5 to Zarr while preserving chunking and compression settings if possible

oruebel commented:
@rly @mavaylon1 can you please  review

rly commented:
The code looks good to me. There are a bunch of branches in the code that translates the compression. I know it's tedious, but could you write tests for those branches? That would help make sure those lines work as intended and prevent our code coverage from dropping 1% to 83.4%. We might want to include hdmf-zarr in our core packages for the U24 at some point.

oruebel commented:
> There are a bunch of branches in the code that translates the compression. I know it's tedious, but could you write tests for those branches?

The main reason I didn't add tests for all the branches is because in order to cover those cases I believe we need to install custom HDF5 compressors, i.e., add `hdf5_plugins` as a dependency just to run tests and `szip` may not be possible to test because it requires a license. In principle, we just need a HDF5 dataset that looks like it uses those filters, but I'm not sure if we can "fake" that easily.

rly commented:
> add hdf5_plugins as a dependency just to run tests

I think that is OK. It would be added to https://github.com/hdmf-dev/hdmf-zarr/blob/dev/requirements-dev.txt

> szip may not be possible to test because it requires a license

We can just exclude that one from coverage

oruebel committed:
Add h5plugins as a dependency for the tests

oruebel committed:
Add unit tests for mapping HDF5 IO filters to Zarr filters via ZarrDataIO

oruebel committed:
Remove test case with system-dependent warning message

oruebel committed:
Add unit test for conversion of byte fillvalue to string

oruebel commented:
@rly Can you please take another look. I added a new unit test module for `ZarrDataIO`. As mentioned above, I had to add `hdf5plugin` as a requirement for these tests. I added ` @unittest.skipIf` to the test cases that use `hdf5plugin` just in case the library is not installed. The tests now cover 100% of the patch. 

<img width="1547" alt="Screen Shot 2024-01-11 at 1 31 26 AM" src="https://github.com/hdmf-dev/hdmf-zarr/assets/10999845/fa52501e-d764-429a-b906-414f3ac99cb6">

rly commented on tests/unit/test_zarrdataio.py:
```diff
@@ -0,0 +1,237 @@
+        # Close the HDF5 file
+        h5file.close()
+
+    def test_from_h5py_datase_bytes_fillvaluet(self):
```
```suggestion
    def test_from_h5py_dataset_bytes_fillvalue(self):
```

rly committed:
Update tests/unit/test_zarrdataio.py

rly commented:
Thanks! Looks good to me.
